xds/internal/xdsclient/authority.go

```diff
@@ -306,365 +325,281 @@ func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Ty
-	if state.cache != nil {
 		if a.logger.V(2) {
-			a.logger.Infof("Resource type %q with resource name %q found in cache: %s", rType.TypeName(), resourceName, state.cache.ToJSON())
+			a.logger.Infof("New watch for type %q, resource name %q", rType.TypeName(), resourceName)
 		}
-		resource := state.cache
-		a.serializer.TrySchedule(func(context.Context) { watcher.OnUpdate(resource, func() {}) })
-	}
 
-	return func() {
-		a.resourcesMu.Lock()
-		defer a.resourcesMu.Unlock()
+		xdsChannel := a.xdsChannelToUseLocked()
+		if xdsChannel == nil {
+			return
+		}
 
-		// We already have a reference to the resourceState for this particular
-		// resource. Avoid indexing into the two-level map to figure this out.
+		// Lookup the entry for the resource type in the top-level map. If there is
+		// no entry for this resource type, create one.
+		resources := a.resources[rType]
+		if resources == nil {
+			resources = make(map[string]*resourceState)
+			a.resources[rType] = resources
+		}
 
-		// Delete this particular watcher from the list of watchers, so that its
-		// callback will not be invoked in the future.
-		state.wState = watchStateCanceled
-		delete(state.watchers, watcher)
-		if len(state.watchers) > 0 {
-			return
+		// Lookup the resource state for the particular resource name that the watch
+		// is being registered for. If this is the first watch for this resource
+		// name, request it from the management server.
+		state := resources[resourceName]
+		if state == nil {
+			if a.logger.V(2) {
+				a.logger.Infof("First watch for type %q, resource name %q", rType.TypeName(), resourceName)
+			}
+			state = &resourceState{
+				watchers:          make(map[xdsresource.ResourceWatcher]bool),
+				md:                xdsresource.UpdateMetadata{Status: xdsresource.ServiceStatusRequested},
+				xdsChannelConfigs: []*xdsChannelWithConfig{xdsChannel},
+			}
+			resources[resourceName] = state
+			xdsChannel.xc.subscribe(rType, resourceName)
 		}
+		// Always add the new watcher to the set of watchers.
+		state.watchers[watcher] = true
 
-		// There are no more watchers for this resource, delete the state
-		// associated with it, and instruct the transport to send a request
-		// which does not include this resource name.
-		if a.logger.V(2) {
-			a.logger.Infof("Removing last watch for type %q, resource name %q", rType.TypeName(), resourceName)
+		// If we have a cached copy of the resource, notify the new watcher.
+		if state.cache != nil {
+			if a.logger.V(2) {
+				a.logger.Infof("Resource type %q with resource name %q found in cache: %s", rType.TypeName(), resourceName, state.cache.ToJSON())
+			}
+			resource := state.cache
+			a.watcherCallbackSerializer.TrySchedule(func(context.Context) { watcher.OnUpdate(resource, func() {}) })
 		}
-		delete(resources, resourceName)
-		a.sendDiscoveryRequestLocked(rType, resources)
-	}
+		cleanup = a.unwatchResource(rType, resourceName, watcher)
+	})
+	<-done
```

Comment:
why we need to spawn a separate thread for watcher if we need to wait for completion anyways?